File: 03-Azure/01-01-App Innovation/05_AppModernizationWithSecurity/java/src/main/java/com/microsoft/microhack/catalog/service/CatalogDocumentParser.java
package com.microsoft.microhack.catalog.service;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;
import com.fasterxml.jackson.core.StreamReadFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.microsoft.microhack.catalog.model.CatalogImportItem;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.regex.Pattern;
import org.springframework.stereotype.Component;

/** Parses and validates complete catalog documents without publishing partial prefixes. */
@Component
public class CatalogDocumentParser {

    private static final Pattern CANONICAL_UUID = Pattern.compile(
            "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$");
    private final ObjectMapper objectMapper;
    private final Validator validator;

    public CatalogDocumentParser(ObjectMapper objectMapper, Validator validator) {
        this.objectMapper = objectMapper.copy()
                .enable(StreamReadFeature.STRICT_DUPLICATE_DETECTION.mappedFeature());
        this.validator = validator;
    }

    /** Returns only after every record and cross-record identity has been validated. */
    public List<ValidatedCatalogItem> parse(InputStream input) {
        List<CatalogImportItem> records = readDocument(input);
        if (records.isEmpty()) {
            throw new CatalogImportValidationException("catalog document must contain at least one record");
        }
        Map<String, String> namesToSlugs = new HashMap<>();
        Map<String, String> slugsToNames = new HashMap<>();
        Set<UUID> productIds = new HashSet<>();
        List<ValidatedCatalogItem> validated = new ArrayList<>(records.size());
        for (CatalogImportItem record : records) {
            ValidatedCatalogItem item = validate(record, namesToSlugs, slugsToNames);
            if (!productIds.add(item.productId())) {
                throw new CatalogImportValidationException(
                        "productId values must be unique within one document");
            }
            validated.add(item);
        }
        return List.copyOf(validated);
    }

    private List<CatalogImportItem> readDocument(InputStream input) {
        try (JsonParser parser = objectMapper.getFactory().createParser(input)) {
            if (parser.nextToken() != JsonToken.START_ARRAY) {
                throw new CatalogImportValidationException(
                        "catalog document root must be one JSON array");
            }
            List<CatalogImportItem> records = new ArrayList<>();
            while (parser.nextToken() != JsonToken.END_ARRAY) {
                if (parser.currentToken() != JsonToken.START_OBJECT) {
                    throw new CatalogImportValidationException(
                            "every catalog array member must be an object");
                }
                records.add(readItem(parser));
            }
            if (parser.nextToken() != null) {
                throw new CatalogImportValidationException(
                        "catalog document must contain exactly one JSON root");
            }
            return records;
        } catch (IOException exception) {
            throw new CatalogImportValidationException("catalog document is not valid JSON", exception);
        } catch (CatalogImportValidationException exception) {
            throw exception;
        } catch (RuntimeException exception) {
            throw new CatalogImportValidationException("catalog document is invalid", exception);
        }
    }

    private CatalogImportItem readItem(JsonParser parser) throws IOException {
        Map<String, String> fields = new HashMap<>();
        while (parser.nextToken() != JsonToken.END_OBJECT) {
            if (parser.currentToken() != JsonToken.FIELD_NAME) {
                throw new CatalogImportValidationException("catalog object contains an invalid token");
            }
            String field = parser.currentName();
            if (!isAllowedField(field)) {
                throw new CatalogImportValidationException("catalog object contains unknown field " + field);
            }
            if (parser.nextToken() != JsonToken.VALUE_STRING) {
                throw new CatalogImportValidationException(
                        "catalog field " + field + " must be a JSON string");
            }
            fields.put(field, parser.getText());
        }
        return new CatalogImportItem(
                required(fields, "productId"),
                required(fields, "name"),
                required(fields, "description"),
                required(fields, "category"),
                required(fields, "filename"),
                required(fields, "imagePrompt"));
    }

    private static boolean isAllowedField(String field) {
        return switch (field) {
            case "productId", "name", "description", "category", "filename", "imagePrompt" -> true;
            default -> false;
        };
    }

    private static String required(Map<String, String> fields, String field) {
        String value = fields.get(field);
        if (value == null) {
            throw new CatalogImportValidationException(
                    "catalog object is missing required field " + field);
        }
        return value;
    }

    private ValidatedCatalogItem validate(
            CatalogImportItem record,
            Map<String, String> namesToSlugs,
            Map<String, String> slugsToNames) {
        validateText(record.name(), "name", 3, 80, null);
        validateText(record.description(), "description", 20, 1200, null);
        validateText(record.category(), "category", 2, 60, null);
        validateText(record.imagePrompt(), "imagePrompt", 30, null, 260);
        Set<ConstraintViolation<CatalogImportItem>> violations = validator.validate(record);
        if (!violations.isEmpty()) {
            throw new CatalogImportValidationException(
                    "catalog record violates " + violations.iterator().next().getPropertyPath());
        }
        if (!CANONICAL_UUID.matcher(record.productId()).matches()) {
            throw new CatalogImportValidationException(
                    "productId must be a canonical lowercase UUID");
        }
        UUID productId;
        try {
            productId = UUID.fromString(record.productId());
        } catch (IllegalArgumentException exception) {
            throw new CatalogImportValidationException("productId is not an RFC 4122 UUID", exception);
        }
        if (!record.productId().equals(productId.toString())) {
            throw new CatalogImportValidationException(
                    "productId must be a canonical lowercase UUID");
        }
        String expectedFilename = record.productId() + ".png";
        if (!expectedFilename.equals(record.filename())) {
            throw new CatalogImportValidationException(
                    "filename must equal productId plus .png");
        }
        String slug = CategorySlug.normalize(record.category());
        if (slug.isEmpty() || slug.length() > 64) {
            throw new CatalogImportValidationException(
                    "category must normalize to a non-empty slug of at most 64 characters");
        }
        String existingSlug = namesToSlugs.putIfAbsent(record.category(), slug);
        String existingName = slugsToNames.putIfAbsent(slug, record.category());
        if ((existingSlug != null && !existingSlug.equals(slug))
                || (existingName != null && !existingName.equals(record.category()))) {
            throw new CatalogImportValidationException(
                    "category display names and slugs must be unique");
        }
        return new ValidatedCatalogItem(
                productId,
                record.name(),
                record.description(),
                record.category(),
                slug,
                record.filename());
    }

    private static void validateText(
            String value,
            String field,
            int minimumCodePoints,
            Integer maximumUtf16Units,
            Integer maximumCodePoints) {
        if (value == null || value.isEmpty() || value.codePoints().allMatch(
                CatalogDocumentParser::isContractWhitespace)) {
            throw new CatalogImportValidationException(field + " must not be blank");
        }
        int codePoints = value.codePointCount(0, value.length());
        if (codePoints < minimumCodePoints) {
            throw new CatalogImportValidationException(
                    field + " must contain at least " + minimumCodePoints + " Unicode code points");
        }
        if (maximumUtf16Units != null && value.length() > maximumUtf16Units) {
            throw new CatalogImportValidationException(
                    field + " must contain at most " + maximumUtf16Units + " UTF-16 code units");
        }
        if (maximumCodePoints != null && codePoints > maximumCodePoints) {
            throw new CatalogImportValidationException(
                    field + " must contain at most " + maximumCodePoints + " Unicode code points");
        }
    }

    private static boolean isContractWhitespace(int codePoint) {
        return Character.isWhitespace(codePoint)
                || Character.isSpaceChar(codePoint)
                || codePoint == 0x0085;
    }
}

File: 03-Azure/01-01-App Innovation/05_AppModernizationWithSecurity/dotnet/src/LegoCatalog.App/Services/CatalogDocumentParser.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LegoCatalog.App.Services;

/// <summary>
/// Parses and validates a complete catalog document before publication starts.
/// </summary>
public sealed class CatalogDocumentParser
{
    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        PropertyNameCaseInsensitive = false,
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Disallow,
    };

    /// <summary>
    /// Returns fully validated records or rejects the entire document.
    /// </summary>
    public async Task<IReadOnlyList<ValidatedCatalogItem>> ParseAsync(
        Stream jsonStream,
        CancellationToken cancellationToken)
    {
        List<CatalogInputItem?>? rawItems;
        try
        {
            rawItems = await JsonSerializer.DeserializeAsync<List<CatalogInputItem?>>(
                jsonStream,
                SerializerOptions,
                cancellationToken);
        }
        catch (JsonException exception)
        {
            throw new CatalogImportValidationException(
                "The catalog document is not valid contract JSON.",
                1,
                exception);
        }

        if (rawItems is null || rawItems.Count == 0)
        {
            throw new CatalogImportValidationException(
                "The catalog document must contain at least one record.",
                1);
        }

        var validated = new List<ValidatedCatalogItem>(rawItems.Count);
        var ids = new HashSet<Guid>();
        for (var index = 0; index < rawItems.Count; index++)
        {
            try
            {
                var item = rawItems[index]
                    ?? throw new CatalogImportValidationException(
                        "catalog records must be JSON objects.",
                        1);
                validated.Add(Validate(item, ids));
            }
            catch (CatalogImportValidationException exception)
            {
                throw new CatalogImportValidationException(
                    $"Record {index + 1} is invalid: {exception.Message}",
                    1,
                    exception);
            }
        }

        return validated;
    }

    private static ValidatedCatalogItem Validate(
        CatalogInputItem item,
        ISet<Guid> ids)
    {
        var name = RequireStoredText(item.Name, 3, 80, "name");
        var description = RequireStoredText(
            item.Description,
            20,
            1200,
            "description");
        var category = RequireStoredText(item.Category, 2, 60, "category");
        RequireCodePointText(item.ImagePrompt, 30, 260, "imagePrompt");
        if (item.ProductId is null
            || !Guid.TryParseExact(item.ProductId, "D", out var productId)
            || !string.Equals(
                productId.ToString("D"),
                item.ProductId,
                StringComparison.Ordinal))
        {
            throw new CatalogImportValidationException(
                "productId must be a canonical lowercase UUID.",
                1);
        }

        if (!ids.Add(productId))
        {
            throw new CatalogImportValidationException(
                "productId values must be unique within one document.",
                1);
        }

        var expectedFilename = $"{productId:D}.png";
        if (!string.Equals(
            item.Filename,
            expectedFilename,
            StringComparison.Ordinal))
        {
            throw new CatalogImportValidationException(
                $"filename must equal {expectedFilename}.",
                1);
        }

        var categorySlug = CategorySlug.Normalize(category);
        if (categorySlug.Length == 0)
        {
            throw new CatalogImportValidationException(
                "category must normalize to a non-empty slug.",
                1);
        }
        if (categorySlug.Length > 64)
        {
            throw new CatalogImportValidationException(
                "category must normalize to at most 64 ASCII characters.",
                1);
        }

        return new ValidatedCatalogItem(
            productId,
            name,
            description,
            category,
            categorySlug,
            item.Filename!);
    }

    private static string RequireStoredText(
        string? value,
        int minimumCodePoints,
        int maximumUtf16CodeUnits,
        string field)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new CatalogImportValidationException(
                $"{field} must not be blank.",
                1);
        }

        if (value.EnumerateRunes().Count() < minimumCodePoints)
        {
            throw new CatalogImportValidationException(
                $"{field} must contain at least {minimumCodePoints} Unicode code points.",
                1);
        }

        if (value.Length > maximumUtf16CodeUnits)
        {
            throw new CatalogImportValidationException(
                $"{field} must contain at most {maximumUtf16CodeUnits} UTF-16 code units.",
                1);
        }

        return value;
    }

    private static string RequireCodePointText(
        string? value,
        int minimum,
        int maximum,
        string field)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new CatalogImportValidationException(
                $"{field} must not be blank.",
                1);
        }

        var length = value.EnumerateRunes().Count();
        if (length < minimum || length > maximum)
        {
            throw new CatalogImportValidationException(
                $"{field} must contain from {minimum} through {maximum} Unicode code points.",
                1);
        }

        return value;
    }

    private sealed record CatalogInputItem(
        [property: JsonPropertyName("productId")] string? ProductId,
        [property: JsonPropertyName("name")] string? Name,
        [property: JsonPropertyName("description")] string? Description,
        [property: JsonPropertyName("category")] string? Category,
        [property: JsonPropertyName("filename")] string? Filename,
        [property: JsonPropertyName("imagePrompt")] string? ImagePrompt);
}

public sealed record ValidatedCatalogItem(
    Guid ProductId,
    string Name,
    string Description,
    string Category,
    string CategorySlug,
    string Filename);

/// <summary>
/// Signals whole-document validation failure before publication.
/// </summary>
public sealed class CatalogImportValidationException : Exception
{
    public CatalogImportValidationException(
        string message,
        int rejectedCount,
        Exception? innerException = null)
        : base(message, innerException)
    {
        RejectedCount = rejectedCount;
    }

    public int RejectedCount { get; }
}
